lib/gems/pending/util/object_storage/miq_ftp_storage.rb

```diff
@@ -0,0 +1,92 @@
+require 'util/miq_ftp_lib'
+require 'util/miq_object_storage'
+require 'logger'
+
+class MiqFtpStorage < MiqObjectStorage
+  include MiqFtpLib
+
+  FTP_CHUNKSIZE = ::Net::FTP::DEFAULT_BLOCKSIZE
+
+  attr_reader :uri, :username, :password
+
+  class << self
+    attr_writer :instance_logger
+
+    def instance_logger
+      @instance_logger ||= $log || Logger.new(File::NULL)
+    end
+  end
+
+  def self.uri_scheme
+    "ftp".freeze
+  end
+
+  def initialize(settings)
+    super
+    @uri      = @settings[:uri]
+    @username = @settings[:username]
+    @password = @settings[:password]
+  end
+
+  # Override for connection handling
```

Comment:
The connection handling in question here is making sure you don't re-open for each file fragment I assume or is there more to it than that?

To be clear I'm fine with the implementation. I'm just curious.

Reply:
Yeah, basically the same reason that we overrode `MiqGenericMountSession#add` as well.  If we aren't splitting, technically it would have only done it once, but with splits, it would do it with each file fragment (as you have mentioned).  `.upload_single` is used in both cases, so overriding there doesn't make sense.

But, possibly it wouldn't hurt to have some follow up documentation (in a different PR) around `MiqFileStorage#add`/`MiqFileStorage#download` stating that those methods can be overwritten for cases like connection handling, but `super` must be called to conform to the spec.